tez-dag/src/main/java/org/apache/tez/dag/app/dag/TaskAttempt.java

```diff
@@ -136,4 +137,5 @@ public void setLocalityCounter(DAGCounter localityCounter) {
    */
   long getFinishTime();
 
+  Set<NodeId> getUnhealthyNodesHistory();
```

Comment:
It would be better to find a better name than "unhealthy". We are really trying to enforce negative affinity for a node, the concept of trying to avoid the same node in case the node is slow for some reason. Many times the original, non-speculative task finishes first. Maybe some better name you can think of.

Reply:
done